onnxruntime/core/graph/graph_utils.cc

```diff
@@ -752,40 +752,51 @@ bool FindPath(Graph& graph, const Node& node, bool is_input_edge, const std::vec
   return true;
 }
 
-bool RemoveNodesWithOneOutputBottomUp(Graph& graph, const Node& start_node) {
-  std::queue<const Node*> q;
-  std::vector<NodeIndex> nodes_to_remove;
-  q.push(&start_node);
+bool RemoveNodesWithOneOutputBottomUp(Graph& graph, const Node& start_node, bool force) {
+  std::queue<NodeIndex> q;
+  std::unordered_set<NodeIndex> removed_nodes;
+  q.push(start_node.Index());
   // From the current node, remove nodes bottom-up util it reaches a node with multiple outputs/graph output.
-  while (q.size() != 0) {
-    const Node& cur_node = *(q.front());
+  while (!q.empty()) {
+    NodeIndex cur_node_index = q.front();
     q.pop();
+
+    if (removed_nodes.find(cur_node_index) != removed_nodes.end()) {
+      continue;
+    }
     // Each eligible node in the subgraph must have less than one output edge and no output should be
     // the graph output
+    const Node& cur_node = *graph.GetNode(cur_node_index);
     if (cur_node.GetOutputEdgesCount() > 1 || !graph.GetNodeOutputsInGraphOutputs(cur_node).empty()) {
       continue;
     }
-    nodes_to_remove.push_back(cur_node.Index());
+
     // push the parents of current node to the queue.
     for (unsigned int i = 0; i < cur_node.InputDefs().size(); ++i) {
       const std::string& input_name = GetNodeInputName(cur_node, i);
       if (IsInitializer(graph, input_name, true) || IsGraphInput(graph, cur_node.InputDefs()[i])) {
         // skip initializers and graph inputs
         continue;
       }
-      q.push(GetInputNode(cur_node, i));
+      const Node* parent_node = GetInputNode(cur_node, i);
+      q.push(parent_node->Index());
+    }
+
+    if (force || cur_node.GetOutputEdgesCount() == 0) {
```

Comment:
force handling shall be moved to the begining of the function like:
       if force:
           remove start node and queue its parents
       else:
           add start node to queue.

       ... handle the queue.

Otherwise, it is not "forced" since the start node might not be removed since there is conditional checks before this line.

Reply:
This is designed intentionally: the first node may have zero, one or multiple outputs. we only want to remove when it has zero or one output.

Reply:
Then update the document about this constraint.